heron/downloaders/src/java/org/apache/heron/downloader/DownloadRunner.java

```diff
@@ -18,17 +18,160 @@
+        .longOpt(CliArgs.HERON_HOME.text)
+        .hasArgs()
+        .argName("heron home dir")
+        .build();
+
+    Option configFile = Option.builder("p")
+        .desc("Path of the config files")
+        .longOpt(CliArgs.CONFIG_PATH.text)
+        .hasArgs()
+        .argName("config path")
+        .build();
+
+    // candidates:
+    // local: download to client local machine
+    // cluster: download into the container in the cloud
+    Option mode = Option.builder("m")
+        .desc("download mode, cluster or local")
+        .longOpt(CliArgs.MODE.text)
+        .hasArg()
+        .argName("download mode")
+        .build();
+
+    options.addOption(packageUri);
+    options.addOption(destination);
+    options.addOption(heronHome);
+    options.addOption(configFile);
+    options.addOption(mode);
+
+    return options;
+  }
+
+
   // takes topology package URI and extracts it to a directory
   public static void main(String[] args) throws Exception {
-    if (args.length != 2) {
-      System.err.println("Usage: downloader <topology-package-uri> <extract-destination>");
+    CommandLineParser parser = new DefaultParser();
+    Options slaManagerCliOptions = constructCliOptions();
+
+    // parse the help options first.
+    Options helpOptions = constructHelpOptions();
+    CommandLine cmd = parser.parse(helpOptions, args, true);
+    if (cmd.hasOption("h")) {
+      usage(slaManagerCliOptions);
       return;
     }
 
-    final String uri = args[0];
-    final String destination = args[1];
+    try {
+      cmd = parser.parse(slaManagerCliOptions, args);
+    } catch (ParseException e) {
+      usage(slaManagerCliOptions);
+      throw new RuntimeException("Error parsing command line options: ", e);
+    }
+
+    DownloaderMode mode = DownloaderMode.cluster;
+    if (cmd.hasOption(CliArgs.MODE.text)) {
+      mode = DownloaderMode.valueOf(cmd.getOptionValue(CliArgs.MODE.text, null));
+    }
+
+    Config config;
+    switch (mode) {
+      case cluster:
+        config = Config.toClusterMode(Config.newBuilder()
+            .putAll(ConfigLoader.loadClusterConfig())
+            .build());
+        break;
+
+      case local:
+        if (!cmd.hasOption(CliArgs.HERON_HOME.text) || !cmd.hasOption(CliArgs.CONFIG_PATH.text)) {
+          throw new IllegalArgumentException("Missing heron_home or config_path argument");
+        }
+        String heronHome = cmd.getOptionValue(CliArgs.HERON_HOME.text, null);
+        String configPath = cmd.getOptionValue(CliArgs.CONFIG_PATH.text, null);
+        config = Config.toLocalMode(Config.newBuilder()
+            .putAll(ConfigLoader.loadConfig(heronHome, configPath, null, null))
+            .build());
+        break;
+
+      default:
+        throw new IllegalArgumentException(
+            "Invalid mode: " + cmd.getOptionValue(CliArgs.MODE.text));
+    }
+
+    final String uri = cmd.getOptionValue(CliArgs.TOPOLOGY_PACKAGE_URI.text, null);
```

Comment:
what happens if uri and destination are null? the old logic has protection and it seems new code doesn't. I think it might be better to have a check here.